### PR TITLE
basic ash drakes

### PR DESCRIPTION
**Pull request by mc-oofert**
## About The Pull Request

https://github.com/tgstation/tgstation/assets/70376633/e50b2a27-b560-4c70-abb1-b42515e7af16

## Why It's Good For The Game

we are trying to make all simple animals basicmobs or somehing

## Changelog
:cl:
refactor: ash drakes are basicmobs
/:cl:

### CODE REVIEW

File: code/modules/mob/living/basic/boss/boss.dm [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+/mob/living/basic/boss/death(gibbed, list/force_grant)
+	if(gibbed) // in case they've been force dusted
+		return ..()
+
+	if(health > 0) // prevents instakills
+		return
+	var/datum/status_effect/crusher_damage/crusher_dmg = has_status_effect(/datum/status_effect/crusher_damage)
+	///Whether we killed the megafauna with primarily crusher damage or not
+	var/crusher_kill = (crusher_dmg && crusher_dmg.total_damage >= maxHealth * 0.6)
+	for(var/loot in (crusher_kill ? crusher_loot : loot)) // might aswell just do it here because the element proves unreliable
```
Comment (Jacquerel) [inline code/modules/mob/living/basic/boss/boss.dm:72]: Fix the element then, if it is broken we want it to work not to go back to recoding the same shit in multiple places.

Comment (mc-oofert) [reply code/modules/mob/living/basic/boss/boss.dm:72]: using two elements is not efficient and having to remove the other whenever someone gets a crusher kill is shit

Comment (Jacquerel) [reply code/modules/mob/living/basic/boss/boss.dm:72]: then we still want a general solution and not one specific to this typepath 🤷 
If you don't want to take the effort to make the code better you're not doing a refactor

**Comment by CCC23** [comment]
please please PLEASE for the love of god make them have a SLIGHT delay between them seeing you and them attacking otherwise its just going to be one miner stumbling upon a ash drake and them immeditly getting combo breakered

**Comment by mc-oofert** [comment]
> please please PLEASE for the love of god make them have a SLIGHT delay between them seeing you and them attacking otherwise its just going to be one miner stumbling upon a ash drake and them immeditly getting combo breakered

idk just kill it

**Comment by LePiromano** [comment]
> idk just kill it

he's true

**Comment by MajManatee** [comment]
Mining was too easy

File: code/modules/mob/living/basic/boss/boss.dm [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+		)
+
+/mob/living/basic/boss/death(gibbed, list/force_grant)
+	if(gibbed) // in case they've been force dusted
+		return ..()
+
+	if(health > 0) // prevents instakills
+		return
+	var/datum/status_effect/crusher_damage/crusher_dmg = has_status_effect(/datum/status_effect/crusher_damage)
+	///Whether we killed the megafauna with primarily crusher damage or not
```
Comment (Jacquerel) [inline code/modules/mob/living/basic/boss/boss.dm:79]:
```suggestion
	// Whether we killed the megafauna with primarily crusher damage or not
```
Triple slash inside a proc messes up the autodocs

File: code/modules/mob/living/basic/boss/boss.dm [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+			adjustBruteLoss(50)
+
+	return TRUE
+
+/// Grants medals and achievements to surrounding players
+/mob/living/basic/boss/proc/grant_achievement(medaltype, scoretype, crusher_kill, list/grant_achievement = list())
+	if(!achievement_type || (flags_1 & ADMIN_SPAWNED_1) || !SSachievements.achievements_enabled) //Don't award medals if the medal type isn't set
+		return FALSE
+	if(!grant_achievement.len)
+		for(var/mob/living/L in view(7,src))
```
Comment (Jacquerel) [inline code/modules/mob/living/basic/boss/boss.dm:156]: One letter variable? L indeed.

File: code/modules/mob/living/basic/boss/drake/drake.dm [hunk 4]
```diff
@@ -277,21 +277,21 @@
...
-
-/obj/effect/temp_visual/target/proc/fall(list/flame_hit)
 	var/turf/T = get_turf(src)
 	playsound(T,'sound/magic/fleshtostone.ogg', 80, TRUE)
 	new /obj/effect/temp_visual/fireball(T)
-	sleep(duration)
+	addtimer(CALLBACK(src, PROC_REF(fall), flame_hit), duration, TIMER_UNIQUE)
+
+/obj/effect/temp_visual/target/proc/fall(list/flame_hit)
+	var/turf/T = get_turf(src)
```
Comment (Jacquerel) [inline code/modules/mob/living/basic/boss/drake/drake.dm:286]: Another one letter var

File: code/modules/mob/living/basic/boss/drake/drake.dm [hunk 5]
```diff
@@ -94,29 +92,31 @@
...
-/mob/living/simple_animal/hostile/megafauna/dragon/Destroy()
+/mob/living/basic/boss/dragon/Destroy()
 	fire_cone = null
 	meteors = null
 	mass_fire = null
 	lava_swoop = null
 	return ..()
 
-/mob/living/simple_animal/hostile/megafauna/dragon/OpenFire()
+/mob/living/basic/boss/dragon/RangedAttack(atom/target)
```
Comment (Jacquerel) [inline code/modules/mob/living/basic/boss/drake/drake.dm:105]: This isn't what you want to hear but all of this should probably be in the AI controlller.
The hint is that you're turning all of this decision-making off when there's a client- that means what we are talking about here is "AI".

**Review by Jacquerel — CHANGES_REQUESTED** [review]

File: code/datums/ai/basic_mobs/basic_subtrees/ranged_skirmish.dm [hunk 6]
```diff
@@ -43,6 +43,11 @@
 	if (distance > max_range || distance < min_range)
 		return AI_BEHAVIOR_DELAY | AI_BEHAVIOR_FAILED
 
+	if(!try_fire(seconds_per_tick, controller, target))
+		return AI_BEHAVIOR_DELAY | AI_BEHAVIOR_FAILED
+
+	return AI_BEHAVIOR_DELAY | AI_BEHAVIOR_SUCCEEDED
+
+/datum/ai_behavior/ranged_skirmish/proc/try_fire(seconds_per_tick, datum/ai_controller/controller, atom/target)
```
Comment (Jacquerel) [inline code/datums/ai/basic_mobs/basic_subtrees/ranged_skirmish.dm:51]: You forgot to make the default version of this proc return anything

File: code/modules/mob/living/basic/boss/drake/drake_ai.dm [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+	attack_behavior = /datum/ai_behavior/ranged_skirmish/drake
+
+/datum/ai_planning_subtree/ranged_skirmish/drake/SelectBehaviors(datum/ai_controller/controller, seconds_per_tick)
+	var/mob/living/target = controller.blackboard[BB_BASIC_MOB_CURRENT_TARGET]
+	if (!istype(target))
+		return
+	if(target.stat != DEAD)
+		return ..()
+
+/datum/ai_behavior/ranged_skirmish/drake
```
Comment (Jacquerel) [inline code/modules/mob/living/basic/boss/drake/drake_ai.dm:32]: I'll be honest I don't know that this has any particularly strong reasons to be a subtype of ranged skirmish
Usually the model we use for AI behaviour is:

Planning subtree- Decide what you actually do
AI behaviour- Do the thing

Probably this should be a planning subtree which performs the probability calculations and then just triggers the "use ability" behaviour for various abilities.

File: code/modules/mob/living/basic/boss/drake/drake_ai.dm [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+	var/datum/action/cooldown/fire_cone = controller.blackboard[BB_DRAKE_FIRECONE]
+	var/datum/action/cooldown/meteors = controller.blackboard[BB_DRAKE_METEORS]
+	if(prob(15 + anger_modifier))
+		if(DRAKE_ENRAGED) // Lava Arena
+			lava_swoop.Trigger(target = target)
+		if(lava_swoop.Trigger(target = target)) // Lava Pools
+			fire_cone.StartCooldown(0)
+			fire_cone.Trigger(target = target)
+			meteors.StartCooldown(0)
+			meteors.Trigger(target = target)
```
Comment (Jacquerel) [inline code/modules/mob/living/basic/boss/drake/drake_ai.dm:53]: Stuff like "using x ability puts y ability on cooldown" also smells very suspicious in an AI behaviour, why do we only want this to happen when they are AI controlled? Should probably be a signal response to this ability being triggered.

**Review by Jacquerel — CHANGES_REQUESTED** [review]

**Comment by mc-oofert** [comment]
fuck basicmobs and uhh i hope nobody downloaded the code

**Comment by LePiromano** [comment]
power move

**Comment by Sakamoto4ka** [comment]
Now show it in 20% avg

**Pull request closed without merging**